src/components/FxParams/validation.ts

```diff
@@ -0,0 +1,91 @@
+import { SafeParseError, SafeParseSuccess, z } from "zod"
+import { FxParamDefinition, FxParamType } from "./types"
+
+const FxParamOptions_numberSchema = z.object({
+  min: z.number().optional(),
+  max: z.number().optional(),
+  step: z.number().optional(),
+})
+
+const FxParamOptions_stringSchema = z.object({
+  minLength: z.number().optional(),
+  maxLength: z.number().optional(),
+})
+
+const FxParamOptions_selectSchema = z.object({
+  options: z.string().array().nonempty(),
+})
+
+export const BaseControllerDefinitionSchema = z.object({
+  id: z.string(),
+  name: z.string().optional(),
+  exposedAsFeature: z.string().optional(),
+})
+
+const StringControllerSchema = BaseControllerDefinitionSchema.extend({
+  type: z.literal("string"),
+  options: FxParamOptions_stringSchema.optional(),
+  default: z.string(),
+})
+
+const NumberControllerSchema = BaseControllerDefinitionSchema.extend({
+  type: z.literal("number"),
+  options: FxParamOptions_numberSchema.optional(),
+  default: z.number(),
+})
+
+const SelectControllerSchema = BaseControllerDefinitionSchema.extend({
+  type: z.literal("select"),
+  options: FxParamOptions_selectSchema.optional(),
+  default: z.string(),
+})
+
+const BooleanControllerSchema = BaseControllerDefinitionSchema.extend({
+  type: z.literal("boolean"),
+  options: z.undefined(),
+  default: z.boolean(),
+})
+
+const ColorControllerSchema = BaseControllerDefinitionSchema.extend({
+  type: z.literal("color"),
+  options: z.undefined(),
+  default: z.string(),
+})
+
+const ControllerDefinitionSchema = z.union([
+  StringControllerSchema,
+  NumberControllerSchema,
+  SelectControllerSchema,
+  BooleanControllerSchema,
+  ColorControllerSchema,
+])
+
+type ControllerDefinitionSchemaType = z.infer<typeof ControllerDefinitionSchema>
+
+export function validateParameterDefinition(
+  parameterDefinition: FxParamDefinition<FxParamType>
+):
+  | SafeParseError<ControllerDefinitionSchemaType>
+  | SafeParseSuccess<ControllerDefinitionSchemaType>
+  | undefined {
+  let schema
```

Comment:
i'm a big fan of this kind of pattern here:

```
const schema = {
  number: NumberControllerSchema,
  color: ColorControllerSchema,
  ...
}[parameterDefinition.type]
```

but definitely just opinionated and not necessary - feel free to ignore!

n.b. for bonus type safety, we can do:
```
const ControllerTypeSchema = z.enum([
  'number',
  'color',
  ...
]);

```

then e.g.: 

```
const ColorControllerSchema = BaseControllerDefinitionSchema.extend({
  type: z.literal(ControllerTypeSchema.enum.color),
  ...
})
```